src/adaXT/decision_tree/decision_tree.pyx

```diff
@@ -0,0 +1,468 @@
+        elif isinstance(max_features, float):
+            return min(num_features, int(max_features * num_features))
+        elif isinstance(max_features, str):
+            if max_features == "sqrt":
+                return int(np.sqrt(num_features))
+            elif max_features == "log2":
+                return int(np.log2(num_features))
+        else:
+            raise ValueError("Unable to parse max_features")
+
+    def __check_input(self, X: object, Y: object):
+        # If the skip_check_input option is set to True in the initialization of a tree, then simply return X and Y
+        if self.skip_check_input:
+            return X, Y
+
+        # Make sure input arrays are c contigous
+        X = np.ascontiguousarray(X, dtype=DOUBLE)
+        Y = np.ascontiguousarray(Y, dtype=DOUBLE)
+
+        # Check if X and Y has same number of rows
+        if X.shape[0] != Y.shape[0]:
+            raise ValueError("X and Y should have the same number of rows")
+
+        # Check if Y has dimensions (n, 1) or (n,)
+        if 2 < Y.ndim:
+            raise ValueError("Y should have dimensions (n,1) or (n,)")
+        elif 2 == Y.ndim:
+            if 1 < Y.shape[1]:
+                raise ValueError("Y should have dimensions (n,1) or (n,)")
+            else:
+                Y = Y.reshape(-1)
+
+        return X, Y
+
+    def __check_sample_weight(self, sample_weight: np.ndarray, n_samples):
+
+        if sample_weight is None:
+            return np.ones(n_samples, dtype=np.double)
+        sample_weight = np.array(sample_weight, dtype=np.double)
+        if sample_weight.shape[0] != n_samples:
+            raise ValueError("sample_weight should have as many elements as X and Y")
+        if sample_weight.ndim > 1:
+            raise ValueError("sample_weight should have dimension (n_samples,)")
+        return sample_weight
+
+    def fit(
+            self,
+            X,
+            Y,
+            sample_indices: np.ndarray | None = None,
+            sample_weight: np.ndarray | None = None,) -> None:
+
+        X, Y = self.__check_input(X, Y)
+        row, col = X.shape
+        self.int_max_features = self.__parse_max_features(self.max_features, col)
+
+        # If sample_weight is valid it is simply passed through check_sample_weight, if it is None all entries are set to 1
+        sample_weight = self.__check_sample_weight(sample_weight=sample_weight, n_samples=row)
+
+        builder = DepthTreeBuilder(
+            X=X,
+            Y=Y,
+            sample_indices=sample_indices,
+            int_max_features=self.int_max_features,
+            sample_weight=sample_weight,
+            criteria=self.criteria_class(X, Y, sample_weight),
+            leaf_builder_class = self.leaf_builder_class,
+            predict_class=self.predict_class,
+            splitter=self.splitter)
+        builder.build_tree(self)
+
+    def predict(self, X: np.ndarray, **kwargs):
+        if not self.root:
```

Comment:
instead of checking for the root, isnt it cleaner to check if the predictor is set?